subprojects/core/src/main/java/org/gradle/api/internal/collections/AbstractIterationOrderRetainingElementSource.java

```diff
@@ -378,6 +384,9 @@ List<T> getValues() {
         }
 
         public boolean remove(T value) {
+            if (removedValues == null) {
```

Comment:
❓ Is this safe because this type implements `WithMutationGuard`?

This question applies to the other two types

Reply:
I think we're trading one sort of unsafeness for another.  None of these classes are thread-safe and the domain object containers are not thread-safe. 

If you look at where this is being called in https://github.com/gradle/gradle/blob/fd9d4ca61ac83f95937ef38b779725aafd22f3e2/subprojects/core/src/main/java/org/gradle/api/internal/collections/AbstractIterationOrderRetainingElementSource.java#L303, you can see this is already deeply broken if used by multiple threads and we have some checks to try to detect/fail early. 

So "Is this thread safe?" - no and it wasn't before.

Maybe it's a bad name, but `WithMutationGuard` is not about thread safety. It's about detecting disallowed configuration. 

This is allowed:
```
tasks.create(...) {
   tasks.register(...)
}
```
because the create is eager.

But this is not:
```
tasks.register(...) {
   tasks.register(...)
}
```
because the outer register is only called when the task is realized.